File: app/src/main/java/djordjeh/architecture/mvp/ui/tasks/TasksPresenter.java
package djordjeh.architecture.mvp.ui.tasks;

import javax.inject.Inject;

import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.data.source.TaskDataSource;
import djordjeh.architecture.mvp.ui.BasePresenterImpl;
import djordjeh.architecture.mvp.ui.SchedulersFacade;

public class TasksPresenter extends BasePresenterImpl<TasksContract.View> implements TasksContract.Presenter {

    private final TaskDataSource taskDataSource;
    private final SchedulersFacade schedulersFacade;

    @Inject
    TasksPresenter(TasksContract.View view, TaskDataSource taskDataSource, SchedulersFacade schedulersFacade) {
        super(view);
        this.taskDataSource = taskDataSource;
        this.schedulersFacade = schedulersFacade;
    }

    @Override
    public void getTasks(boolean forceUpdate) {
        addDisposable(
                taskDataSource.tasks(forceUpdate)
                        .subscribeOn(schedulersFacade.io())
                        .observeOn(schedulersFacade.mainThread())
                        .subscribe(view::showTasks, view::showError, () -> view.showProgress(false))
        );
    }

    @Override
    public void saveTask(Task task) {
        addDisposable(taskDataSource.save(task)
                .subscribeOn(schedulersFacade.io())
                .observeOn(schedulersFacade.mainThread())
                .subscribe(view::showSavedMessage, view::showError));
    }

    @Override
    public void deleteTask(Task task) {
        addDisposable(taskDataSource.delete(task)
                .subscribeOn(schedulersFacade.io())
                .observeOn(schedulersFacade.mainThread())
                .subscribe(view::showTaskDeletedMessage, view::showError));
    }

    @Override
    public void addNewTask() {
        view.showTask(null);
    }

    @Override
    public void showTask(Task task) {
        view.showTask(task);
    }

    @Override
    public void onRefresh() {
        getTasks(true);
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/ui/tasks/TasksPresenter.kt
package djordjeh.architecture.mvp.ui.tasks

import javax.inject.Inject

import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.data.source.TaskDataSource
import djordjeh.architecture.mvp.ui.BasePresenterImpl
import djordjeh.architecture.mvp.ui.SchedulersFacade

class TasksPresenter @Inject
internal constructor(
        view: TasksContract.View,
        private val taskDataSource: TaskDataSource,
        private val schedulersFacade: SchedulersFacade) : BasePresenterImpl<TasksContract.View>(view), TasksContract.Presenter {

    override fun getTasks(forceUpdate: Boolean) {
        addDisposable(
                taskDataSource.tasks(forceUpdate)
                        .subscribeOn(schedulersFacade.io())
                        .observeOn(schedulersFacade.mainThread())
                        .subscribe( { view.showTasks(it) }, { view.showError(it) }, { view.showProgress(false) })
        )
    }

    override fun saveTask(task: Task) {
        addDisposable(taskDataSource.save(task)
                .subscribeOn(schedulersFacade.io())
                .observeOn(schedulersFacade.mainThread())
                .subscribe({ view.showSavedMessage(it) }, { view.showError(it) }))
    }

    override fun deleteTask(task: Task) {
        addDisposable(taskDataSource.delete(task)
                .subscribeOn(schedulersFacade.io())
                .observeOn(schedulersFacade.mainThread())
                .subscribe({ view.showTaskDeletedMessage(it) }, { view.showError(it) }))
    }

    override fun addNewTask() {
        view.showTask(null)
    }

    override fun showTask(task: Task) {
        view.showTask(task)
    }

    override fun onRefresh() {
        getTasks(true)
    }
}
